xdsl/dialects/experimental/aie.py

```diff
@@ -0,0 +1,1129 @@
+                    "length": length,
+                    "AB": AB,
+                    "dimensions": dimensions,
+                },
+                operands=[buffer],
+            )
+
+    def print(self, printer: Printer):
+        printer.print("(<")
+        printer.print_operand(self.buffer)
+        printer.print(
+            ": ",
+            self.buffer.type,
+            ", ",
+            self.offset.value.data,
+            ", ",
+            self.length.value.data,
+            ">, ",
+            self.ab.value.data,
+            ")",
+        )
+
+
+@irdl_op_definition
+class DMABDPACKETOp(IRDLOperation):
+    name = "AIE.dmaBdPacket"
+    packet_type = attr_def(IntegerAttr[IntegerType])
+    packet_id = attr_def(IntegerAttr[IntegerType])
+
+    def __init__(
+        self, packet_type: IntegerAttr[IntegerType], packet_id: IntegerAttr[IntegerType]
+    ):
+        super().__init__(
+            attributes={"packet_type": packet_type, "packet_id": packet_id}
+        )
+
+    def print(self, printer: Printer):
+        printer.print(
+            "(",
+            f"0x{self.packet_type.value.data:X}",
+            ", ",
+            f"0x{self.packet_id.value.data:X}",
+            ")",
+        )
+
+
+# TODO: add successor basic blocks
+@irdl_op_definition
+class DMAStartOp(IRDLOperation):
+    name = "AIE.dmaStart"
+    channelDir = attr_def(IntegerAttr[Annotated[IntegerType, i32]])
+    channelIndex = attr_def(IntegerAttr[IntegerType])
+    dest = successor_def()
+    chain = successor_def()
+
+    traits = frozenset([IsTerminator()])
+
+    def __init__(
+        self,
+        channelDir: IntegerAttr[IntegerType],
+        channelIndex: IntegerAttr[IntegerType],
+        dest: Block,
+        chain: Block,
+    ):
+        super().__init__(
+            attributes={"channelDir": channelDir, "channelIndex": channelIndex},
+            successors=[dest, chain],
+        )
+
+    def print(self, printer: Printer):
+        direction = "MM2S" if self.channelDir.value.data == 0 else "S2MM"
+        printer.print("(", direction, ", ", self.channelIndex.value.data, ", ")
+        printer.print_block_name(self.dest)
+        printer.print(", ")
+        printer.print_block_name(self.chain)
+        printer.print(")")
+
+
+@irdl_op_definition
+class DebugOp(IRDLOperation):
+    name = "AIE.debug"
+    arg = operand_def(AnyAttr())
+
+    def __init__(self, arg: Operation | SSAValue):
+        super().__init__(operands=[arg])
+
+
+@irdl_op_definition
+class DeviceOp(IRDLOperation):
+    name = "AIE.device"
+
+    region = region_def()
+
+    device = attr_def(IntegerAttr[Annotated[IntegerType, i32]])
+    traits = frozenset([SymbolTable()])
+
+    def __init__(self, device: IntegerAttr[IntegerType], region: Region):
+        super().__init__(attributes={"device": device}, regions=[region])
+
+    def print(self, printer: Printer):
+        printer.print("(")
+        device_str = "xcvc1902" if self.device.value.data == 0 else ""
```

Comment:
Something about this feels off to me, I'd expect the textual IR to reflect the value in the attribute. I have a vague recollection of @PapyChacal doing something conceptually related with enums but I might be wrong

Reply:
Completely missed that ping!
I implemented EnumAttribute, but it wouldn't help for printing a keyword-prefixed integer?
The declarative assembly format soon should have support for that.
That's the job of optional groups, that would be written 
```
( `xcvc1902` $device^ )?
```
And attribute directives and optional groups are both open on separate PRs :slightly_smiling_face:

Reply:
That statement makes use of an EnumAttribute for the type of device. I'm afraid the declarative assembly format is not an option here because this operation has a region. Unless this capability has been added over the last week.